app/flatpak-builtins-update.c

```diff
@@ -67,51 +67,72 @@ static GOptionEntry options[] = {
 
   if (remote == NULL)
     {
       g_auto(GStrv) remotes = NULL;
-      int i;
 
-      remotes = flatpak_dir_list_remotes (dir, cancellable, error);
-      if (remotes == NULL)
-        return FALSE;
-
-      for (i = 0; remotes[i] != NULL; i++)
+      for (j = 0; j < dirs->len; j++)
         {
-          g_autoptr(GError) local_error = NULL;
-          FlatpakTerminalProgress terminal_progress = { 0 };
-
-          if (flatpak_dir_get_remote_disabled (dir, remotes[i]) ||
-              flatpak_dir_get_remote_noenumerate (dir, remotes[i]))
-            continue;
-
-          g_print (_("Updating appstream for remote %s\n"), remotes[i]);
-          g_autoptr(OstreeAsyncProgress) progress = flatpak_progress_new (flatpak_terminal_progress_cb, &terminal_progress);
-          if (!flatpak_dir_update_appstream (dir, remotes[i], opt_arch, &changed,
-                                             progress, cancellable, &local_error))
-            g_printerr (_("Error updating: %s\n"), local_error->message);
-          ostree_async_progress_finish (progress);
-          flatpak_terminal_progress_end (&terminal_progress);
+          FlatpakDir *dir = g_ptr_array_index (dirs, j);
+
+          remotes = flatpak_dir_list_remotes (dir, cancellable, error);
+          if (remotes == NULL)
+            return FALSE;
+
+          for (i = 0; remotes[i] != NULL; i++)
+            {
+              g_autoptr(GError) local_error = NULL;
+              FlatpakTerminalProgress terminal_progress = { 0 };
+
+              if (flatpak_dir_get_remote_disabled (dir, remotes[i]) ||
+                  flatpak_dir_get_remote_noenumerate (dir, remotes[i]))
+                continue;
+
+              g_print (_("Updating appstream for remote %s\n"), remotes[i]);
+              g_autoptr(OstreeAsyncProgress) progress = flatpak_progress_new (flatpak_terminal_progress_cb, &terminal_progress);
+              if (!flatpak_dir_update_appstream (dir, remotes[i], opt_arch, &changed,
+                                                 progress, cancellable, &local_error))
+                g_printerr (_("Error updating: %s\n"), local_error->message);
+              ostree_async_progress_finish (progress);
+              flatpak_terminal_progress_end (&terminal_progress);
+            }
         }
     }
   else
     {
-      FlatpakTerminalProgress terminal_progress = { 0 };
-      g_autoptr(OstreeAsyncProgress) progress = flatpak_progress_new (flatpak_terminal_progress_cb, &terminal_progress);
-      res = flatpak_dir_update_appstream (dir, remote, opt_arch, &changed,
-                                          progress, cancellable, error);
-      ostree_async_progress_finish (progress);
-      if (!res)
-        return FALSE;
-      flatpak_terminal_progress_end (&terminal_progress);
+      gboolean found = FALSE;
+
+      for (j = 0; j < dirs->len; j++)
+        {
+          FlatpakDir *dir = g_ptr_array_index (dirs, j);
+
+          if (flatpak_dir_has_remote (dir, remote))
```

Comment:
Wouldn't it make sense to use `flatpak_resolve_duplicate_remotes` like the other commands do?

Reply:
The new behaviour in general in this commit is that if you specify no --user or --system, then all matching things are updated. For instance, flatpak update org.gnome.gedit will update both the system and the user gedit (if installed), and we error only if the specifice ref doesn't exist in any of the dirs. Similarly, flatpak update --appstream updates appstream data for all remotes in both system and user.

So, the question is, should flatpak update --appstream flathub update appstream in both user and system to match the above behaviour, or should it stop at the user appstream?

I chose to match the other update behaviour, given that an update (especially of appdata) is a pretty "safe" operation, whereas accidentally not updating when you think it did can be a pain.

I guess the overall question is whether the "update both system and user" is the right behaviour for update though. Opinions?